src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/json/EventTransformer.scala

```diff
@@ -67,23 +67,95 @@ object EventTransformer {
-    if (event.length != Fields.size) {
+    getValidatedJsonEvent(event)
+  }
+
+  /**
+    * Converts an array of field values to a validated JSON whose keys are the field names corresponding to the
+    * EnrichedEvent POJO of the Scala Common Enrich project. If there are any self-describing events in the fields:
+    * "unstruct_event", "derived_contexts" or "contexts" these are returned in a "shredded" format (e.g.
+    * "unstruct_event_com_acme_1_myField": "value"). This default behaviour can also be disabled, so that the deep
+    * structure is returned for those three fields instead, by setting the flatten parameter to false.
+    *
+    * @param event Array of values for the event
+    * @param flatten Whether to flattent the fields in "unstruct_event", "contexts" and "derived_contexts"
+    * @return ValidatedRecord containing JSON for the event and the event_id (if it exists)
+    */
+  def getValidatedJsonEvent(event: Array[String], flatten:Boolean = true): Validated[(Set[InventoryItem], JObject)] = {
+    if (isRightSize(event)) {
       Left(List(s"Expected ${Fields.size} fields, received ${event.length} fields. This may be caused by attempting to use this SDK version on an older (pre-R73) or newer version of Snowplow enriched events."))
     } else {
-      val geoLocation: JObject = {
-        val latitude = event(GeopointIndexes.latitude)
-        val longitude = event(GeopointIndexes.longitude)
-        if (latitude.nonEmpty && longitude.nonEmpty) {
-          JObject("geo_location" -> JString(s"$latitude,$longitude"))
-        } else {
-          JObject()
-        }
-      }
+      val geoLocation: JObject = getGeoLocationField(event)
+      if (flatten) convertEvent(event.toList, geoLocation)
+      else getJObjectWithNestedStructures(event)
+    }
+  }
+
+  private def getGeoLocationField(event: Array[String]): JObject = {
+    val latitude = event(GeopointIndexes.latitude)
+    val longitude = event(GeopointIndexes.longitude)
+    if (latitude.nonEmpty && longitude.nonEmpty) {
+      JObject("geo_location" -> JString(s"$latitude,$longitude"))
+    } else {
+      JObject()
+    }
+  }
+
+  private def getJObjectWithNestedStructures(event: Array[String]): Validated[(Set[InventoryItem], JObject)] = {
+    for {
+      jObject <- convertTsvEventToNestedJObject(event)
+    } yield (Set.empty[InventoryItem], getGeoLocationField(event) ~ jObject)
+  }
 
-      convertEvent(event.toList, geoLocation)
+  /**
+    * Event has been checked for size
+    *
+    * @param event the event array
+    * @return either a list of errors or the JObject representation
+    */
+  private def convertTsvEventToNestedJObject(event: Array[String]): Validated[JObject]= {
+    val validJObjects = FieldNames.zip(event).zip(FieldTypes).map {
+      case ((fieldName: String, fieldValue: String), fieldType: TsvToJsonConverter) => fieldParser(fieldName, fieldValue, fieldType)
```

Comment:
Why type annotations are necessary here? It makes function look partial.

Reply:
For Debugging